File: ROS/mw/mw_video/src/image_converter.cpp
#include <ros/ros.h>
#include <image_transport/image_transport.h>
#include <cv_bridge/cv_bridge.h>
#include <sensor_msgs/image_encodings.h>
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <sys/time.h>

#define FRAME_HEIGHT 360
#define FRAME_WIDTH  640

class ImageConverter {
  ros::NodeHandle nh_;
  image_transport::ImageTransport it_;
  image_transport::Subscriber image_sub_;
  image_transport::Publisher image_pub_;
  double distance;
  double delta_heading;
  long long time_micros;
  int count;

public:
  ImageConverter() : it_(nh_) {
    time_micros = 0;
    count = 0;
    // Subscrive to input video feed and publish output video feed
    image_sub_ = it_.subscribe("/camera/rgb/image_raw", 1, 
      &ImageConverter::imageCb, this);
    image_pub_ = it_.advertise("/image_converter/output_video", 1);
  }

  ~ImageConverter() {
  }

  void imageCb(const sensor_msgs::ImageConstPtr& msg) {
    cv_bridge::CvImagePtr cv_ptr;
    cv_bridge::CvImagePtr processed_ptr;

    count++;

    if (count % 100 == 0) {
      long long current_micros;
      struct timeval tv;

      gettimeofday(&tv, NULL);
      current_micros = (long long)tv.tv_sec * 1000000 + tv.tv_usec;
      printf("time: %f\n", (double)(current_micros  - time_micros) / 100);
      time_micros = current_micros;
    }

    // Use cv_bridge() to convert the ROS image to OpenCV format
    try {
      cv_ptr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::BGR8);
    }
    catch (cv_bridge::Exception& e) {
      ROS_ERROR("cv_bridge exception: %s", e.what());
      return;
    }

    // Process the frame using the process_image() function
    processed_ptr = process_image(cv_ptr);
                       
    // Output modified video stream
    image_pub_.publish(processed_ptr->toImageMsg());
  }

  cv_bridge::CvImagePtr process_image(cv_bridge::CvImagePtr cv_ptr) {
    cv::Mat blured;
    cv::Mat hsv;

    cv::blur(cv_ptr->image, blured, cv::Size(3, 3));
    cv::cvtColor(blured, hsv, CV_RGB2HSV);

    // works kind of thresholded = cv2.inRange(hsv, cv.Scalar(0, 190, 125, 0), cv.Scalar(15, 255, 230, 0))
    //               thresholded2 = cv2.inRange(hsv, cv.Scalar(165, 190, 125, 0), cv.Scalar(180, 255, 230, 0))
    // thresholded = cv2.inRange(hsv, (0, 210, 125, 0), (10, 255, 255, 0))
    // thresholded2 = cv2.inRange(hsv, (170, 210, 125, 0), (180, 255, 255, 0))

    cv::Mat thresholded, thresholded2;

    cv::inRange(hsv, cv::Scalar(0, 210, 125, 0), cv::Scalar(25, 255, 255, 0), thresholded);
    cv::inRange(hsv, cv::Scalar(155, 210, 125, 0), cv::Scalar(180, 255, 255, 0), thresholded2);
    cv::bitwise_or(thresholded, thresholded2, thresholded);

    std::vector<std::vector<cv::Point> > contours;
    std::vector<cv::Vec4i> hierarchy;
    int max_index = -1;
    double a_max = -1.0;

    cv::findContours(thresholded, contours, hierarchy, CV_RETR_TREE, CV_CHAIN_APPROX_SIMPLE, cv::Point(0, 0));

    for (int i = 0; i < contours.size(); i++ ) {
      double a = cv::contourArea(cv::Mat(contours[i]));

      if (a > a_max) {
        a_max = a;
        max_index = i;
      }
    }

    if (max_index >= 0) {
      std::vector<cv::Point> cnt = contours[max_index];
      std::vector<cv::Point> poly;
      cv::Rect r;

      cv::approxPolyDP(cv::Mat(cnt), poly, 3, true);
      r = cv::boundingRect(cv::Mat(poly));

      if (r.width * r.height > 160 && r.width / r.height < 1.0) {
        std::vector<std::vector<cv::Point> > poly_vec;

        poly_vec.push_back(poly);
        distance += 0.24 * FRAME_HEIGHT / (0.3969 * r.height);
        delta_heading += (70.42 / FRAME_WIDTH) * ((r.x + r.width / 2) - (FRAME_WIDTH / 2));
        cv::drawContours(cv_ptr->image, poly_vec, 0, cv::Scalar(12, 236, 160), 1);
        cv::rectangle(cv_ptr->image, r.tl(), r.br(), cv::Scalar(12, 236, 160), 2);
      }
    }

    return cv_ptr;
  }
};

int main(int argc, char** argv) {
  ros::init(argc, argv, "image_converter");
  ImageConverter ic;
  ros::spin();
  return 0;
}

File: ROS/mw/mw_video/nodes/image_converter.py
#!/usr/bin/env python

import cv2
import math
import numpy as np
import sys
import time

import rospy
from sensor_msgs.msg import Image, CameraInfo
from cv_bridge import CvBridge, CvBridgeError
from mw_video.msg import ConeLocation

class cvBridgeFindCone():
    def __init__(self):
        self.node_name = "mwFindCone"
        self.time_micros = time.time()
        self.processing_time = 0
        self.average_processing_time = 0
        self.average_time = 0
        self.count = 0
        self.distance = 0
        self.delta_heading = 0
        self.found = False
        # self.frame_height = 480
        # self.frame_width = 864
        self.frame_height = 360
        self.frame_width = 640

        rospy.init_node(self.node_name)
        
        # What we do during shutdown
        rospy.on_shutdown(self.cleanup)

        self.is_debug = rospy.get_param('~debug', False)
        rospy.loginfo('%s: Parameter %s has value %s', self.node_name, rospy.resolve_name('~debug'), self.is_debug)

        # Create the cv_bridge object
        self.bridge = CvBridge()

        self.image_sub = rospy.Subscriber("/camera/rgb/image_raw", Image, self.image_callback, queue_size = 1)
        self.cone_pub = rospy.Publisher("/mw/cone_location", ConeLocation, queue_size = 1)
        self.image_pub = None

        if self.is_debug:
            self.image_pub = rospy.Publisher("/mw/image_topic", Image, queue_size = 1)

        rospy.loginfo("%s: Ready." % self.node_name)

    def image_callback(self, ros_image):
        # Use cv_bridge() to convert the ROS image to OpenCV format
        self.count = self.count + 1
        frame = None

        try:
            frame = self.bridge.imgmsg_to_cv2(ros_image, "bgr8")
        except CvBridgeError, e:
            print e
        
        if frame == None:
            return

        # Process the frame using the process_image() function
        t = time.time()
        display_image = self.process_image(frame)
        self.processing_time += time.time() - t

        if self.is_debug: 
            try:
                self.image_pub.publish(self.bridge.cv2_to_imgmsg(display_image, "bgr8"))
            except CvBridgeError, e:
                print e

        if self.found:
            self.found = False
            rospy.loginfo("%s: distance: %f  delta_heading: %f" % (self.node_name, self.distance, self.delta_heading))
            cl = ConeLocation(heading = self.delta_heading, distance = self.distance)
            self.cone_pub.publish(cl)

        if self.count % 100 == 0:
            current_micros = time.time()
            self.average_time =(current_micros  - self.time_micros) / 100
            self.average_processing_time = self.processing_time / 100
            self.processing_time = 0
            rospy.loginfo("%s: time: %f  %f\n" % (self.node_name, self.average_time, self.average_processing_time))
            self.time_micros = current_micros

    def process_image(self, frame):
        blurMat = cv2.blur(frame, (3,3))
        hsv = cv2.cvtColor(blurMat, cv2.COLOR_BGR2HSV)

        # thresholded = cv2.inRange(hsv, (0, 190, 125, 0), (15, 255, 230, 0))
        # thresholded2 = cv2.inRange(hsv, (165, 190, 125, 0), (180, 255, 230, 0))
        # 20150402 thresholded = cv2.inRange(hsv, (0, 210, 125, 0), (10, 255, 255, 0))
        # 20150402 thresholded2 = cv2.inRange(hsv, (170, 210, 125, 0), (180, 255, 255, 0))
        thresholded = cv2.inRange(hsv, (0, 170, 125, 0), (10, 255, 255, 0))
        thresholded2 = cv2.inRange(hsv, (170, 170, 125, 0), (180, 255, 255, 0))
        # thresholded = cv2.inRange(hsv, (0, 210, 125, 0), (25, 255, 255, 0))
        # thresholded2 = cv2.inRange(hsv, (155, 210, 125, 0), (180, 255, 255, 0))

        thresholded = cv2.bitwise_or(thresholded, thresholded2)

        ret = cv2.findContours(thresholded, cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)
        contours = None

        if len(ret) > 2:
            contours = ret[1]
        elif len(ret) == 2:
            contours = ret[0]

        if self.is_debug:
            t = "Avg Time: %6.4f (%6.4f)" % (self.average_time, self.average_processing_time)

        if contours != None:
            areas = [cv2.contourArea(c) for c in contours]

            if areas and len(areas) > 0:
                max_index = np.argmax(areas)
                cnt = contours[max_index]

                poly = cv2.approxPolyDP(cnt, 3, True)
                x,y,w,h =  cv2.boundingRect(poly)

                if w * h > 160 and w / h < 1.0:
                    self.distance = 0.24 * self.frame_height / (0.3969 * h)
                    self.delta_heading = (70.42 / self.frame_width) * ((x + w / 2) - (self.frame_width / 2))
                    self.found = True

                    if self.is_debug:
                        cv2.drawContours(frame, [poly], 0, (12, 236, 160), 1)
                        cv2.rectangle(frame, (x, y), (x + w, y + h), (12, 236, 160), 2)
                        t += " Dist: %6.2fm   Head: %6.2f" % (self.distance, self.delta_heading)

        if self.is_debug:
            cv2.putText(frame, t, (10, 350), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 60, 245), 2)

        return frame
    
    def cleanup(self):
        print "Shutting down vision node."
    
def main(args):       
    try:
        cvBridgeFindCone()
        rospy.spin()
    except KeyboardInterrupt:
        print "close"

if __name__ == '__main__':
    main(sys.argv)
    
